src/transparency/orig_proto.rs

```diff
@@ -0,0 +1,244 @@
+        if let Some(orig_proto) = req.headers().get(L5D_ORIG_PROTO) {
+            trace!("detected orig-proto: {:?}", orig_proto);
+            let val = orig_proto.as_bytes();
+            let was_absolute_form = was_absolute_form(val);
+            let host = bind::Host::detect(req);
+
+            return bind::Protocol::Http1 {
+                host,
+                is_h1_upgrade: false, // orig-proto is never used with upgrades
+                was_absolute_form,
+            };
+        }
+    }
+
+    bind::Protocol::detect(req)
+}
+
+
+// ===== impl Upgrade =====
+
+impl<S> Upgrade<S> {
+    pub fn new(inner: S, upgrade_h1: bool) -> Self {
+        Self {
+            inner,
+            upgrade_h1,
+        }
+    }
+}
+
+impl<S, B1, B2> Service for Upgrade<S>
+where
+    S: Service<Request = http::Request<B1>, Response = http::Response<B2>>,
+{
+    type Request = S::Request;
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = future::Map<
+        S::Future,
+        fn(S::Response) -> S::Response
+    >;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut req: Self::Request) -> Self::Future {
+        let mut downgrade_response = false;
+
+        if self.upgrade_h1 && req.version() != http::Version::HTTP_2 {
+            debug!("upgrading {:?} to HTTP2 with orig-proto", req.version());
+
+            // absolute-form is far less common, origin-form is the usual,
+            // so only encode the extra information if it's different than
+            // the normal.
+            let was_absolute_form = h1::is_absolute_form(req.uri());
+            if !was_absolute_form {
+                // Since the version is going to set to HTTP_2, the NormalizeUri
+                // middleware won't normalize the URI automatically, so it
+                // needs to be done now.
+                h1::normalize_our_view_of_uri(&mut req);
+            }
+
+            let val = match (req.version(), was_absolute_form) {
+                (http::Version::HTTP_11, false) => "HTTP/1.1",
+                (http::Version::HTTP_11, true) => "HTTP/1.1; absolute-form",
+                (http::Version::HTTP_10, false) => "HTTP/1.0",
+                (http::Version::HTTP_10, true) => "HTTP/1.0; absolute-form",
+                (v, _) => unreachable!("bad orig-proto version: {:?}", v),
+            };
+            req.headers_mut().insert(
+                L5D_ORIG_PROTO,
+                HeaderValue::from_static(val)
+            );
+
+
+            *req.version_mut() = http::Version::HTTP_2;
+            downgrade_response = true;
+        }
+
+        let fut = self.inner.call(req);
+
+        if downgrade_response {
+            fut.map(|mut res| {
+                debug_assert_eq!(res.version(), http::Version::HTTP_2);
+                let version = if let Some(orig_proto) = res.headers().get(L5D_ORIG_PROTO) {
```

Comment:
Just to confirm... it's necessary to use a response header because the request and response may advertise different HTTP versions.

Reply:
Yep exactly. The request may say its `HTTP/1.1`, but if the response comes back as `HTTP/1.0`, we should be sure to pass that along (and vice versa).